internal/config/seccomp/seccomp.go

```diff
@@ -29,50 +31,106 @@ var (
 			Names: []string{
-				unshareName,
+				"clone",
+				"clone3",
+				"unshare",
 			},
 			Action: seccomp.ActAllow,
 			Includes: seccomp.Filter{
 				Caps: []string{"CAP_SYS_ADMIN"},
 			},
 		})
+
+		var flagsIndex uint = 0
+		if runtime.GOARCH == "s390" || runtime.GOARCH == "s390x" {
+			flagsIndex = 1
+		}
+
+		prof.Syscalls = append(prof.Syscalls,
+			&seccomp.Syscall{
+				Name:   "clone",
+				Action: seccomp.ActAllow,
+				Args: []*seccomp.Arg{
+					{
+						Index:    flagsIndex,
+						Value:    unix.CLONE_NEWNS | unix.CLONE_NEWUTS | unix.CLONE_NEWIPC | unix.CLONE_NEWUSER | unix.CLONE_NEWPID | unix.CLONE_NEWNET | unix.CLONE_NEWCGROUP,
+						ValueTwo: 0,
+						Op:       seccomp.OpMaskedEqual,
+					},
+				},
+			},
+			&seccomp.Syscall{
+				Name:   "clone",
+				Action: seccomp.ActErrno,
+				Errno:  "EPERM",
+				Args: []*seccomp.Arg{
+					{Index: flagsIndex, Value: unix.CLONE_NEWNS, ValueTwo: unix.CLONE_NEWNS, Op: seccomp.OpMaskedEqual},
+					{Index: flagsIndex, Value: unix.CLONE_NEWUTS, ValueTwo: unix.CLONE_NEWUTS, Op: seccomp.OpMaskedEqual},
+					{Index: flagsIndex, Value: unix.CLONE_NEWIPC, ValueTwo: unix.CLONE_NEWIPC, Op: seccomp.OpMaskedEqual},
+					{Index: flagsIndex, Value: unix.CLONE_NEWUSER, ValueTwo: unix.CLONE_NEWUSER, Op: seccomp.OpMaskedEqual},
+					{Index: flagsIndex, Value: unix.CLONE_NEWPID, ValueTwo: unix.CLONE_NEWPID, Op: seccomp.OpMaskedEqual},
+					{Index: flagsIndex, Value: unix.CLONE_NEWNET, ValueTwo: unix.CLONE_NEWNET, Op: seccomp.OpMaskedEqual},
+					{Index: flagsIndex, Value: unix.CLONE_NEWCGROUP, ValueTwo: unix.CLONE_NEWCGROUP, Op: seccomp.OpMaskedEqual},
+				},
+				Excludes: seccomp.Filter{
+					Caps: []string{"CAP_SYS_ADMIN"},
+				},
+			},
+			// Because seccomp currently can't compare the data inside struct and the flags in clone3 are hidden in a struct,
+			// seccomp can't block clone3 based on its flags. To force it to use only clone, we make clone3 return ENOSYS,
+			// so that glibc can fall back to clone in the same way as https://github.com/moby/moby/pull/42681.
+			&seccomp.Syscall{
+				Name:   "clone3",
+				Action: seccomp.ActErrno,
+				Errno:  "ENOSYS",
+				Excludes: seccomp.Filter{
+					Caps: []string{"CAP_SYS_ADMIN"},
+				},
+			})
 		defaultProfile = prof
 	})
 
 	return defaultProfile
 }
 
+// validateSyscallIndex checks if the syscall's index matches the default profile's index.
+// We know the default profile at compile time, though a vendor change may update it.
+// Panic on error and have CI catch errors on vendor bumps to avoid combing through.
+func validateSyscallIndex(prof *seccomp.Seccomp, name string, parentStructIndex, index int) {
+	if prof.Syscalls[parentStructIndex].Names[index] == name {
+		return
+	}
+	i := slices.Index(prof.Syscalls[parentStructIndex].Names, name)
+	if i == -1 {
+		logrus.Fatalf("Default seccomp profile updated and syscall moved. Change the parentStructIndex for %q", name)
+	}
```

Comment:
```suggestion
	if i := slices.Index(prof.Syscalls[parentStructIndex].Names, name); i == -1 {
		logrus.Fatalf("Default seccomp profile updated and syscall moved. Change the parentStructIndex for %q", name)
	}
```

Reply:
So, when the user sees this error message and CRI-O exists, what ought they do? What does it mean to "change the parentStructIndex"?

Reply:
yeah maybe this should be "contact the CRI-O maintainers if you see this" sort of message.

Reply:
This is not a message for users. This error is supposed to be caught during CI, at least `ctr_seccomp.bats` should catch it if the upstream default is changed.

https://github.com/cri-o/cri-o/blob/6b41d86972a240ed62663fb984a1b829a2837357/internal/config/seccomp/seccomp.go#L119-L121

Reply:
You mean the users are the contributors?
The cri-o users can't see this error because this function always passes if CI passes.
This function just checks the discrepancies of the indexes [1] and default seccomp [2], and doesn't take any user configurations.

However the contributors to CRI-O will see this error message when they upgrade seccomp package (they can't ignore it though because CI fails), and the message can be improved as you say.

[1] https://github.com/cri-o/cri-o/blob/6b41d86972a240ed62663fb984a1b829a2837357/internal/config/seccomp/seccomp.go#L39-L47

[2] https://github.com/cri-o/cri-o/blob/a46e1d98ac29bd6088184266a0a7c2f9592cd90f/vendor/github.com/containers/common/pkg/seccomp/default_linux.go#L50-L58

Reply:
How about something like this?
```go
	if i == -1 {
		logrus.Errorf("Change the ParentStructIndex for %q", name)
	} else {
		logrus.Errorf("Change the Index for %q to %d", name, i)
	}
	logrus.Fatalf(
		`Default seccomp has been changed, and cri-o couldn't remove some risky syscalls.
You are seeing this error because you updated "github.com/containers/common/pkg/seccomp".
Please look up "vendor/github.com/containers/common/pkg/seccomp/default_linux.go,"
and change DefaultProfile() in internal/config/seccomp/seccomp.go according to the message above.`,
		)
```

Reply:
> The cri-o users can't see this error because this function always passes if CI passes.

If someone compiles it themselves and potentially does something like changing how the module vendoring works they could see it.

> `Default seccomp has been changed, and cri-o couldn't remove some risky syscalls. You are seeing this error because you updated "github.com/containers/common/pkg/seccomp". Please look up "vendor/github.com/containers/common/pkg/seccomp/default_linux.go," and change DefaultProfile() in internal/config/seccomp/seccomp.go according to the message above.`,

I think that error assumes too much about the situation. Something like this is accurate but gives some leeway for various things going wrong and doesn't accuse the user themselves of changing something:

```
`The default internal seccomp policy has been changed and cri-o can't adjust some risky syscalls. You are likely seeing this error because "github.com/containers/common/pkg/seccomp" was updated. Please contact the developers or change DefaultProfile() in internal/config/seccomp/seccomp.go to match according to the message above.`,
```